server/src/test/java/io/spine/server/aggregate/IdempotencyGuardShould.java

```diff
@@ -0,0 +1,149 @@
+
+        repository = new TestAggregateRepository();
+        boundedContext.register(repository);
+    }
+    
+    @After
+    public void tearDown() throws Exception {
+        repository.close();
+        boundedContext.close();
+    }
+
+    @Test(expected = DuplicateCommandException.class)
+    public void throw_DuplicateCommandException_when_the_command_was_handled_since_last_snapshot() {
+        final TenantId tenantId = newTenantId();
+        final ProjectId projectId = newProjectId();
+        final Command createCommand = command(createProject(projectId), tenantId);
+
+        final CommandBus commandBus = boundedContext.getCommandBus();
+        final StreamObserver<Ack> noOpObserver = noOpObserver();
+        commandBus.post(createCommand, noOpObserver);
+
+        final TestAggregate aggregate = getAggregate(projectId, tenantId);
+        final IdempotencyGuard guard = new IdempotencyGuard(aggregate);
+        guard.check(of(createCommand));
+    }
+
+    @Test
+    public void not_throw_when_the_command_was_handled_but_the_snapshot_was_made() {
+        repository.setSnapshotTrigger(1);
+
+        final TenantId tenantId = newTenantId();
+        final ProjectId projectId = newProjectId();
+        final Command createCommand = command(createProject(projectId), tenantId);
+
+        final CommandBus commandBus = boundedContext.getCommandBus();
+        final StreamObserver<Ack> noOpObserver = noOpObserver();
+        commandBus.post(createCommand, noOpObserver);
+
+        final TestAggregate aggregate = getAggregate(projectId, tenantId);
+
+        final IdempotencyGuard guard = new IdempotencyGuard(aggregate);
+        guard.check(of(createCommand));
+    }
+
+    @Test
+    public void not_throw_if_the_command_was_not_handled() {
+        final TenantId tenantId = newTenantId();
+        final ProjectId projectId = newProjectId();
+        final Command createCommand = command(createProject(projectId), tenantId);
+        final TestAggregate aggregate = new TestAggregate(projectId);
+
+        final IdempotencyGuard guard = new IdempotencyGuard(aggregate);
+        guard.check(of(createCommand));
+    }
+
+    @Test
+    public void not_throw_if_another_command_was_handled() {
+        final TenantId tenantId = newTenantId();
+        final ProjectId projectId = newProjectId();
+        final Command createCommand = command(createProject(projectId), tenantId);
+        final Command startCommand = command(startProject(projectId), tenantId);
+
+        final CommandBus commandBus = boundedContext.getCommandBus();
+        final StreamObserver<Ack> noOpObserver = noOpObserver();
+        commandBus.post(createCommand, noOpObserver);
+
+        final TestAggregate aggregate = getAggregate(projectId, tenantId);
+
+        final IdempotencyGuard guard = new IdempotencyGuard(aggregate);
+        guard.check(of(startCommand));
+    }
+
+    private TestAggregate getAggregate(ProjectId id, TenantId tenantId) {
```

Comment:
This seems to be `loadAggregate`.

Reply:
This was actually a piece duplicated from another test environment.

@mdrachuk, please see `AbstractAggregateTestRepository`, which eliminates the duplication.